examples/linear_algebra.dx

```diff
@@ -0,0 +1,196 @@
+      sRef!i := (b.i - s) / a.i.(0@_) -- 0 is the diagonal index
+
+-- Todo: get rid of these by writing a dependent indexing (!) operator.
+def lowerTriIndex (ref:Ref h (LowerTriMat n)) (i:n) : Ref h ((..i)=>Float) =
+  %indexRef ref i
+def upperTriIndex (ref:Ref h (UpperTriMat n)) (i:n) : Ref h ((i..)=>Float) =
+  %indexRef ref i
+
+
+'### Permutations
+
+def Permutation (n:Type) : Type = n=>n
+def apply_permutation (permutation: n=>n) (array: n=>t) : n=>t =
+  for i. array.(permutation.i)
+def identity_permutation (n:Type) ?-> : Permutation n =
+  for i. i
+
+
+'### LU decomposition functions
+
+def pivotize (_:Eq n) ?=> (a:n=>n=>Float) : (Permutation n & Int) =
+  -- Rearranges a matrix to make Gaussian elimination more stable.
+  -- Returns permutation and number of swaps in permutation,
+  -- for use in computing the determinant.
+  snd $ withState (identity_permutation, 0) \stateRef.
+    (pRef, countRef) = (fstRef stateRef, sndRef stateRef)
+    for j:n.
+      row_with_largest = argmin for i:(j..). (-(abs a.(%inject i).j))
+      row_with_largest = %inject row_with_largest
+      case (j == row_with_largest) of
+        True -> ()
+        False -> 
+          tempj = get pRef!j  -- Is there a refSwap?
+          pRef!j := get pRef!row_with_largest
+          pRef!row_with_largest := tempj
+          countRef := (get countRef) + 1  -- TODO: Use += 1
+
+def permutation_sign (swapcount:Int) : Float =
+  -- Computes the sign of the determinant of a permutation
+  -- given the number of row swaps.
+  case mod swapcount 2 == 0 of
+    True -> 1.0
+    False -> -1.0
+
+def lu (_:Eq n) ?=> (a: n=>n=>Float) :
+       (LowerTriMat n & UpperTriMat n & Permutation n & Int) =
+  -- Computes lower, upper, and permuntation matrices from a square matrix,
+  -- such that apply_permutation permutation a == lower ** upper.
+  -- I.e. it uses partial pivoting.
+  (permutation, swapcount) = pivotize a
+  a = apply_permutation permutation a
+
+  init_lower = for i:n. for j':(..i).
+    select (i == (%inject j')) 1.0 0.0
+  init_upper = for i:n. for j'':(i..). 0.0
+
+  (lower, upper) = snd $ withState (init_lower, init_upper) \stateRef.
+    lRef = fstRef stateRef
+    uRef = sndRef stateRef
+
+  -- For reference, here's code to computed the LU decomposition
+  -- without dependent tables (i.e. with standard flat matrices):
+  --  for j:n.
+  --    for i:(..j).
+  --      i = %inject i
+  --      s = sum for k':(..i).
+  --        k = %inject k'
+  --        (get uRef!k!j) * (get lRef!i!k)
+  --      uRef!i!j := a.i.j - s
+
+  --    for i':(j<..).
+  --      i = %inject i'
+  --      s = sum for k':(..j).
+  --        k = %inject k'
+  --        (get uRef!k!j) * (get lRef!i!k)
+  --      lRef!i!j := (a.i.j - s) / (get uRef!j!j)
+  --    for i:n. ()
+
+    for j:n.
+      for i:(..j).
+        i' = %inject i
+        s = sum for k:(..i).
+          k' = (ordinal k)@n
+          ukj = get (upperTriIndex uRef k')!(((ordinal j) - (ordinal k))@_)
```

Comment:
This part seems to diverge from the reference above? There, you do `uRef!k!j` not `uRef!k!(j - k)` (with subtraction happening on ordinals).

Reply:
Oh is it because it's upper triangular?